rfc/20240513-static-evaluation-providers.md

```diff
@@ -99,12 +94,18 @@ locals {
 
 module "mod" {
   source = "./mod"
-  providers {
-    aws = aws[local.region] # Uses the new reference format.
+  providers = {
+    # The new instance key segment can use arbitrary expressions in the index brackets.
+    aws = aws.by_region[local.region]
   }
 }
 ```
 
+The `provider` argument for `resource`/`data` blocks and the `providers` argument for `module` blocks retains much of the rigid static structure previously required, to ensure that it remains possible to statically recognize the relationships between resource configuration blocks and provider configuration blocks since that would be required for a later fully-dynamic version of this feature implemented in the main language runtime: it must be able to construct the dependency graph before evaluating any expressions.
+
+However, the new "index" portion delimited by square brackets is, from a syntax perspective, allowed to contain any valid HCL expression. The only constraints on that expression are on what it is derived from and on its result: it must be derived only from values known during planning, its result must be a value that convert to the type `string`, and the string after conversion must match one of the instance keys of the indicated provider configuration.
```

Comment:
"it must be derived only from values known during planning".  I think for now, it makes more sense to have this known during configuration due to how the existing transformers function.

Reply:
I must admit that I wrote this part trying to make sense of something that the existing implementation was already doing, but perhaps I misunderstood what it was doing.

I had assumed that the index _must_ be dynamic because otherwise we wouldn't be able to refer to `each.key` unless the resource or module `for_each` were also static eval. Forcing resource and module `for_each` to be static-eval compatible _only when there happens to be a dynamic expression in the `provider` argument_ seems like a very subtle, hard-to-underrstand rule.

Please let me know if I misunderstood what you were suggesting.

Reply:
That is the way it's currently implemented, and your understanding is correct.  We do have a pretty reasonable error for when this occurs, but it is an odd edge that users can run into.

The main reason is that we want to be able to enumerate all of the provider "instances" that might be used during the initial TransformProviders.  Given the additional restrictions mentioned in this PR, I think we could probably defer the actual instance lookup to later on in the graph and instead depend on all of the providers that could be referred to by a given module/type/alias combo.

I think that for the sake of timing, we could keep this restriction in the first alpha and potentially remove it in the beta.  It would be broadening the scope of what's allowed by this feature definition.

Reply:
Okay! I will study the current partial implementation again to make sure I properly understand exactly what the restriction is, and then document it here.

Reply:
@cam72cam and I subsequently spoke synchronously and made a different plan:

We're going to try to see what it would take to make it behave how I described, where the instance key expression is evaluated dynamically. It seems like that might actually be simpler overall but does require some changes to work already done.

If that experiment is successful then we'll run with it behaving like I wrote it here. If it seems too complicated to change in the first round then we'll leave it with the static eval approach for first release and then consider making it dynamic in a later change. If we decide to defer it then I'll update the RFC to describe the updated plan.